File: translated/f-to-c/DRB096-doall2-taskloop-collapse-orig-no.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//Two-dimensional array computation:
//Two loops are associated with omp taskloop due to collapse(2).
//Both loop index variables are private.
//taskloop requires OpenMP 4.5 compilers. No data race pairs.

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

// Global variable (from module)
int** a;

int main()
{
    int len, i, j;
    
    len = 100;

    // Allocate 2D array
    a = (int**)malloc(len * sizeof(int*));
    for (i = 0; i < len; i++) {
        a[i] = (int*)malloc(len * sizeof(int));
    }

    #pragma omp parallel
    {
        #pragma omp single
        {
            #pragma omp taskloop collapse(2)
            for (i = 0; i < len; i++) {
                for (j = 0; j < len; j++) {  // Both i and j are private due to collapse(2)
                    a[i][j] = a[i][j] + 1;
                }
            }
        }
    }

    printf("a(50,50) = %d\n", a[49][49]);  // Adjust for 0-based indexing

    // Free 2D array
    for (i = 0; i < len; i++) {
        free(a[i]);
    }
    free(a);

    return 0;
} 

File: translated/f-to-go/DRB096-doall2-taskloop-collapse-orig-no.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//Two-dimensional array computation:
//Two loops are associated with omp taskloop due to collapse(2).
//Both loop index variables are private.
//taskloop requires OpenMP 4.5 compilers. No data race pairs.

package main

import (
	"fmt"
	"runtime"
	"sync"
)

// Package-level variable (module equivalent)
var a [][]int

func main() {
	var length int
	length = 100

	a = make([][]int, length)
	for i := 0; i < length; i++ {
		a[i] = make([]int, length)
	}

	//$omp parallel
	var wg sync.WaitGroup
	numCPU := runtime.NumCPU()

	for threadID := 0; threadID < numCPU; threadID++ {
		wg.Add(1)
		go func() {
			defer wg.Done()

			//$omp single
			if threadID == 0 { // Only one thread executes single region
				//$omp taskloop collapse(2)
				// collapse(2) flattens nested loops for task distribution
				var taskWg sync.WaitGroup
				totalIterations := length * length

				for iteration := 0; iteration < totalIterations; iteration++ {
					taskWg.Add(1)
					go func(iteration int) { // iteration variables are private to each task
						defer taskWg.Done()
						// Convert flat iteration back to 2D indices
						i := (iteration / length) + 1 // Fortran 1-based
						j := (iteration % length) + 1 // Fortran 1-based
						a[i-1][j-1] = a[i-1][j-1] + 1 // No race - proper task partitioning
					}(iteration)
				}
				taskWg.Wait()
				//$omp end taskloop
			}
			//$omp end single
		}()
	}
	wg.Wait()
	//$omp end parallel

	fmt.Printf("a(50,50) = %3d\n", a[49][49])
}
